File: Java/Leetcode/src/Medium/SearchInRotatedSortedArray.java
package Medium;

public class SearchInRotatedSortedArray {

    public int search(int[] nums, int target) {
        int start = 0; // 왼쪽
        int end = nums.length - 1; // 오른쪽
        int mid; // 중간

        // 이분 탐색 시작
        while(start <= end){
            mid = (start + end) / 2;
            if(nums[mid] == target) return mid; // 값 찾으면 return

            // start - mid가 순서대로 정렬 되었을때
            else if(nums[start] <= nums[mid]){
                // start - mid 사이에 target 값이 있을 경우
                if(nums[start] <= target && target < nums[mid]){
                    end = mid - 1;
                }else{
                    start = mid + 1;
                }
                // mid - end 순서대로 정렬 되었을때
            }else{
                //  mid - end 사이 값이 있을 경우
                if(nums[mid] < target && target <= nums[end]){
                    start = mid + 1;
                }else{
                    end = mid - 1;
                }
            }
        }

        return -1;
    }
    public static void main(String[] args) {
        SearchInRotatedSortedArray solution = new SearchInRotatedSortedArray();
        int[] nums = {4,5,6,7,0,1,2};
//        System.out.println(solution.search(nums, 0));
//        System.out.println(solution.search(nums, 3));
        nums = new int[] {1, 3};
//        System.out.println(solution.search(nums, 3));
        nums = new int[] {5, 1, 3};
//        System.out.println(solution.search(nums, 5));
        nums = new int[] {4,5,6,7,8,1,2,3};
        System.out.println(solution.search(nums, 8));
    }
}

File: Python/Leetcode/Medium/Search in Rotated Sorted Array.py
class Solution:
    def search(self, nums: list[int], target: int) -> int:
        left = 0 # 왼쪽
        right = len(nums) - 1 # 오른쪽

        # 이진탐색 시작
        while left <= right:
            mid = (left + right) // 2 # 중간값 계산
            # 찾았을때 위치 반환
            if target == nums[mid]:
                return mid

            # 왼쪽 - 중간이 정렬 되어 있을때
            elif nums[left] <= nums[mid]:
                # 왼쪽 - 중간안에 target이 있을 경우
                if nums[left] <= target < nums[mid]:
                    right = mid - 1
                else:
                    left = mid + 1
            # 중간 - 오른쪽이 정렬 되어 있을때
            else:
                # 중간 - 오른쪽 사이에 값이 존재 할때
                if nums[mid] < target <= nums[right]:
                    left = mid + 1
                else:
                    right = mid - 1
        return -1


solution = Solution()
nums = [4, 5, 6, 7, 8, 1, 2, 3]
print(solution.search(nums, 8))
